modules/openid_connect/app/services/openid_connect/user_tokens/fetch_service.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) 2012-2024 the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module OpenIDConnect
+  module UserTokens
+    class FetchService
+      include Dry::Monads[:result]
+
+      def initialize(user:, jwt_parser: JwtParser.new(verify_audience: false, verify_expiration: false))
+        # TODO: fail if no provider? fail if user is not using OIDC?
+        # or optimistically assume that if there is a token, everything will be alright?
+        @user = user
+        @provider = user.authentication_provider
```

Comment:
I am going to remove the TODO comment above and choose the "optimistic" path, unless anyone stops me.

Rationale: For the `provider` to be missing, the user would either not be using OIDC or the provider would have to be deleted.

In the former case, the user should not have any `oidc_user_tokens`, so there is no way that we'd ever try to use the provider to refresh a token.

In the latter case, there would be no way for the user to log-in.